tests/phpunit/api/v3/JobTest.php

```diff
@@ -320,16 +320,124 @@ public function testBatchMergeWorks($dataSet) {
+    $this->contributionCreate(array('contact_id' => $contactID));
+    $this->contributionCreate(array('contact_id' => $contact2ID, 'invoice_id' => '2', 'trxn_id' => 2));
+    $this->contactMembershipCreate(array('contact_id' => $contactID));
+    $this->contactMembershipCreate(array('contact_id' => $contact2ID));
+    $this->activityCreate(array('source_contact_id' => $contactID, 'target_contact_id' => $contactID, 'assignee_contact_id' => $contactID));
+    $this->activityCreate(array('source_contact_id' => $contact2ID, 'target_contact_id' => $contact2ID, 'assignee_contact_id' => $contact2ID));
+    $this->tagCreate(array('name' => 'Tall'));
+    $this->tagCreate(array('name' => 'Short'));
+    $this->entityTagAdd(array('contact_id' => $contactID, 'tag_id' => 'Tall'));
+    $this->entityTagAdd(array('contact_id' => $contact2ID, 'tag_id' => 'Short'));
+    $this->entityTagAdd(array('contact_id' => $contact2ID, 'tag_id' => 'Tall'));
+    $result = $this->callAPISuccess('Job', 'process_batch_merge', array('mode' => 'safe'));
+    $this->assertEquals(0, count($result['values']['skipped']));
+    $this->assertEquals(1, count($result['values']['merged']));
+    $this->callAPISuccessGetCount('Contribution', array('contact_id' => $contactID), 2);
+    $this->callAPISuccessGetCount('Contribution', array('contact_id' => $contact2ID), 0);
+    $this->callAPISuccessGetCount('FinancialItem', array('contact_id' => $contactID), 2);
+    $this->callAPISuccessGetCount('FinancialItem', array('contact_id' => $contact2ID), 0);
+    $this->callAPISuccessGetCount('Membership', array('contact_id' => $contactID), 2);
+    $this->callAPISuccessGetCount('Membership', array('contact_id' => $contact2ID), 0);
+    $this->callAPISuccessGetCount('EntityTag', array('contact_id' => $contactID), 2);
+    $this->callAPISuccessGetCount('EntityTag', array('contact_id' => $contact2ID), 0);
+    // 12 activities is one for each contribution (2), one for each membership (+2 = 4)
+    // 3 for each of the added activities as there are 3 roles (+6 = 10
+    // 2 for the (source & target) contact merged activity (+2 = 12)
+    $this->callAPISuccessGetCount('ActivityContact', array('contact_id' => $contactID), 12);
+    // 2 for the connection to the deleted by merge activity (source & target)
+    $this->callAPISuccessGetCount('ActivityContact', array('contact_id' => $contact2ID), 2);
+  }
+
+  /**
+   * Check that the merge carries across various related entities.
+   *
+   * Note the group combinations 'expected' results:
+   *
+   * Group 0  Added  null  Added
+   * Group 1  Added  Added  Added
+   * Group 2  Added  Removed  ****  Added
+   * Group 3  Removed  null  **** null
+   * Group 4  Removed  Added  **** Added
+   * Group 5  Removed  Removed **** null
+   * Group 6  null  Added  Added
+   * Group 7  null  Removed  **** null
+   *
+   * The ones with **** are the ones where I think a case could be made to change the behaviour.
+   */
```

Comment:
Good thinking, definitely worth the test, and I think we should JIRA it. Although it's data loss I don't think it's critical. But we should fix it up.